src/gibbs.jl

```diff
@@ -0,0 +1,225 @@
+        start_idx = end_idx + 1
+    end
+
+    # ensure the order of the keys is the same as the one in variable_sizes
+    return NamedTuple{Tuple(keys(variable_sizes))}([
+        result[name] for name in keys(variable_sizes)
+    ])
+end
+
+"""
+    update_trace(trace::NamedTuple, gibbs_state::GibbsState)
+
+Update the trace with the values from the MCMC states of the sub-problems.
+"""
+function update_trace(trace::NamedTuple, gibbs_state::GibbsState)
+    for parameter_variable in keys(gibbs_state.mcmc_states)
+        sub_state = gibbs_state.mcmc_states[parameter_variable]
+        trace = merge(
+            trace,
+            unflatten(
+                vec(sub_state),
+                NamedTuple{(parameter_variable,)}((
+                    gibbs_state.variable_sizes[parameter_variable],
+                )),
+            ),
+        )
+    end
+    return trace
+end
+
+function AbstractMCMC.step(
+    rng::Random.AbstractRNG,
+    logdensity_model::AbstractMCMC.LogDensityModel,
+    sampler::Gibbs,
+    args...;
+    initial_params::NamedTuple,
+    kwargs...,
+)
+    if Set(keys(initial_params)) != Set(sampler.parameter_names)
+        throw(
+            ArgumentError(
+                "initial_params must contain all parameters in the model, expected $(sampler.parameter_names), got $(keys(initial_params))",
+            ),
+        )
+    end
+
+    mcmc_states = Dict{Symbol,Any}()
+    variable_sizes = Dict{Symbol,Tuple}()
+    for parameter_variable in sampler.parameter_names
+        sub_sampler = sampler.sampler_map[parameter_variable]
+
+        variables_to_be_conditioned_on = setdiff(
+            sampler.parameter_names, (parameter_variable,)
+        )
+        conditioning_variables_values = NamedTuple{Tuple(variables_to_be_conditioned_on)}(
+            Tuple([initial_params[g] for g in variables_to_be_conditioned_on])
+        )
+        sub_problem_parameters_values = NamedTuple{(parameter_variable,)}((
+            initial_params[parameter_variable],
+        ))
+
+        # LogDensityProblems' `logdensity` function expects a single vector of real numbers
+        # `Gibbs` stores the parameters as a named tuple, thus we need to flatten the sub_problem_parameters_values
+        # and unflatten after the sampling step
+        variable_sizes[parameter_variable] = Tuple(size(initial_params[parameter_variable]))
+        flattened_sub_problem_parameters_values = flatten(sub_problem_parameters_values)
+
+        sub_state = last(
+            AbstractMCMC.step(
+                rng,
+                AbstractMCMC.LogDensityModel(
+                    AbstractMCMC.condition(
+                        logdensity_model.logdensity, conditioning_variables_values
+                    ),
+                ),
+                sub_sampler,
+                args...;
+                initial_params=flattened_sub_problem_parameters_values,
+                kwargs...,
+            ),
+        )
```

Comment:
As as above, much prefer the following

```suggestion
        sub_model = AbstractMCMC.LogDensityModel(
            AbstractMCMC.condition(
                logdensity_model.logdensity, conditioning_variables_values
            ),
        )
        sub_state = last(
            AbstractMCMC.step(
                rng,
                sub_model,
                sub_sampler,
                args...;
                initial_params=flattened_sub_problem_parameters_values,
                kwargs...,
            ),
        )
```